src/components/toggleable-code-block/toggleable-code-block.js

```diff
@@ -1,52 +1,227 @@
+    const {
+      highlightedCode,
+      limitHeight,
+      copyRanges,
+      selectedLanguage
+    } = this.props;
+    const snippetProps = {
+      code: code,
+      highlightedCode: highlightedCode,
+      maxHeight: limitHeight ? 480 : undefined,
+      highlightThemeCss: highlightTheme,
+      onCopy: () => {}
+    };
+    if (copyRanges) {
+      snippetProps.copyRanges = copyRanges[selectedLanguage];
+      return <NumberedCodeSnippet {...snippetProps} />;
+    } else {
+      return <CodeSnippet {...snippetProps} />;
     }
-    this.state = {
-      toggleValue: prefLanguage
+  };
+
+  renderTitle = () => {
+    const { filename, link } = this.props;
+    const titleProps = {
+      filename: filename,
+      link: link ? link : undefined
     };
-  }
+    return <CodeSnippetTitle {...titleProps} />;
+  };
+
+  renderToggle = () => {
+    const { id, changeLanguage, options } = this.props;
+    return (
+      <div className="mb12">
+        <CodeToggle
+          id={id}
+          onChange={value => {
+            changeLanguage(value);
+          }}
+          options={options}
+        />
+      </div>
+    );
+  };
+
+  checkPreference = language => {
+    const { selectedLanguage } = this.props;
+    return selectedLanguage === language;
+  };
 
   render() {
-    const { props } = this;
-    const currentCodeSnippet = props.codeSnippet.filter(snippet => {
-      return snippet.preferredLanguage === true;
-    })[0];
-    const renderedCodeBlock = (
-      <CodeSnippet
-        style={{ background: '#273d56' }}
-        code={currentCodeSnippet.rawCode}
-        highlightedCode={currentCodeSnippet.highlightedCode}
-        maxHeight={480}
-        onCopy={() => {}}
-      />
+    const { filename, options, code } = this.props;
+
+    return (
+      <div className="my24">
+        {filename && this.renderTitle()}
+        {options && options.length > 1 ? this.renderToggle() : ''}
+        {code && this.renderSnippet(code)}
+      </div>
     );
-    return <div>{renderedCodeBlock}</div>;
   }
 }
 
 ToggleableCodeBlock.propTypes = {
-  codeSnippet: PropTypes.arrayOf(
-    PropTypes.shape({
-      language: PropTypes.oneOf([
-        'swift',
-        'objective-c',
-        'java',
-        'kotlin',
-        'javascript'
-      ]),
-      rawCode: PropTypes.string.isRequired,
-      highlightedCode: PropTypes.string.isRequired,
-      preferredLanguage: PropTypes.bool.isRequired
-    })
-  ).isRequired
+  /* A unique `id` is required for the language toggle. */
+  id: PropTypes.string.isRequired,
+  /*  */
+  selectedLanguage: PropTypes.string.isRequired,
+  /* Raw (unhighlighted) code. When the user clicks a copy
+  button, this is what they'll get. If no highlightedCode
+  is provided, code is displayed. */
+  code: PropTypes.string.isRequired,
+  /* The HTML output of running code through a syntax highlighter.
+  This should use a /helper/higlight-* function. */
+  highlightedCode: PropTypes.string.isRequired,
```

Comment:
The previous comment on line 205 makes it sound like highlightedCode is optional - should it be required?

Reply:
Yea it's not required for the mr-ui component, but I'd like us to always use it in the toggleable code block component. What do you think? I can update the comment above.

Reply:
I think making it required makes sense!